### ISSUE TITLE
Removing huge number of aliases make the cluster unstable

### ISSUE THREAD

**Issue by mdojwa** [elastic/elasticsearch#15186]
ES version 1.4.4

I have 250K aliases in my ES cluster pointing the same index. Removing huge number (over 1000) aliases in 1 batch request (https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html) makes the whole cluster unstable. Master node uses 100% CPU and a lot of search and indexing requests ends up with timeout at client side. I am not sure if these are requests that query/index data from shards placed on master node only or from any node, but a lot of requests to any node time out.

Once removing aliases query finishes everything gets back to normal.

**Comment by jpountz** [comment]
> Master node uses 100% CPU

Interesting. Any chance that you can capture [hot threads](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-hot-threads.html) on the master node to see where this cpu goes?

**Comment by mdojwa** [comment]
Hi, sure.
Here https://gist.github.com/mdojwa/a54d3c97738836952f83 you will find 5 hot_threads dumps done sequentially while the request deleting aliases was running. The master node was es6-1

I was asked to create this ticket by ES support team. They said: "...When ES tries to resolve what it needs to do, for each command, it ends up looping over all the aliases each time."

**Comment by ywelsch** [comment]
Which API are you using?

Currently, having an aliases request of the form
```
POST /_aliases
{
    "actions" : [
        { "remove" : { "index" : "test1", "alias" : "alias1" } },
        { "remove" : { "index" : "test1", "alias" : "alias2" } },
        { "remove" : { "index" : "test1", "alias" : "alias3" } },
        { "remove" : { "index" : "test1", "alias" : "alias4" } },
       ...
    ]
}
```
indeed leads to the issue that the alias name is resolved against all alias names for the index for each "remove" entry.

Would be nice if you could specify

```
POST /_aliases
{
    "actions" : [
        { "remove" : { "index" : "test1", "alias" : "alias1,alias2,alias3,alias4,..." } },
       ...
    ]
}
```

but that's currently not possible.

Only the `_alias` endpoint provides this at the moment:

```
/{index}/_alias/alias1,alias2,alias3,...
```
With that though you might run into request header size limits.

**Comment by clintongormley** [comment]
Good thinking @ywelsch

**Comment by bleskes** [comment]
agreed this would be the simplest now (as opposed to be smarted about executing arbitrary individual actions). Can we also support an array format?

```
POST /_aliases
{
    "actions" : [
        { "remove" : { "index" : "test1", "alias" : [ "alias1" ,"alias2" ... ] } },
       ...
    ]
}
```

**Comment by ywelsch** [comment]
I have implemented @bleskes's suggestion in #15305 by adding an array syntax for aliases (as well as indices). A minor difference to his suggestion is that I used a different field name ("aliases" instead of "alias") for the array syntax so that the types (simple value vs array) are not confused. The old syntax is still valid.

As an example, the following will remove multiple aliases in one go:

```
POST /_aliases
{
    "actions" : [
        { "remove" : { "index" : "test1", "aliases" : [ "alias1" ,"alias2" ... ] } },
       ...
    ]
}
```

Note that, using the Java client, specifying multiple aliases was already syntactically possible, but forbidden by a later check. This check has been removed in #15305.
